### ISSUE TITLE
nn::loss::SmoothL1Loss is detached from the autograd graph (Phase 6 spike 002)

### ISSUE THREAD

**Issue by guyernest** [paiml/aprender#3034]
## Summary

`aprender::nn::loss::SmoothL1Loss::forward` computes its loss with a raw `Vec<f32>` map and
then constructs a **fresh** `Tensor` from that data. The result is DETACHED from the autograd
graph: backward through it contributes nothing, so any model trained with this loss silently
receives zero gradient from it.

Found while porting NeuralProphet (Phase 6, spike 002). The port could not use core's loss and
had to build its own from differentiable ops.

## The code (`crates/aprender-core/src/nn/loss.rs:170-195`)

```rust
    #[must_use]
    pub fn forward(&self, pred: &Tensor, target: &Tensor) -> Tensor {
        assert_eq!(pred.shape(), target.shape());

        let diff = pred.sub(target);
        let loss_data: Vec<f32> = diff
            .data()
            .iter()
            .map(|&x| {
                let abs_x = x.abs();
                if abs_x < self.beta {
                    0.5 * x * x / self.beta
                } else {
                    abs_x - 0.5 * self.beta
                }
            })
            .collect();

        let loss = Tensor::new(&loss_data, pred.shape());   // <-- graph ends here

        match self.reduction {
            Reduction::None => loss,
            Reduction::Mean => loss.mean(),
            Reduction::Sum => loss.sum(),
        }
    }
```

`diff` is on the graph. `diff.data()` leaves it. `Tensor::new(&loss_data, ..)` starts a new,
parentless tensor, so the `.mean()` / `.sum()` that follows has nothing upstream to reach.

## Workaround in use today

`aprender_forecast::np::weighted_huber` (`crates/aprender-forecast/src/np.rs`) implements the
same loss purely from differentiable tensor ops and stays connected. It is asserted connected
by `np::tests::weighted_huber_is_graph_connected`, and the equation
`huber_built_from_ops` in `contracts/neuralprophet-parity-v1.yaml` exists precisely because
of this defect. That workaround lives in a forecasting crate and is not a fix for core.

## Proposed fix

Compose the loss from ops rather than from a data map, so the graph is never left:

- `abs_diff = diff.abs()`
- quadratic branch: `diff.pow(2.0).mul_scalar(0.5 / beta)`
- linear branch: `abs_diff.sub_scalar(0.5 * beta)`
- select with a CONSTANT mask derived from `abs_diff < beta` — the mask is data-dependent but
  not differentiable, so building it from raw data is correct; the two branches must not be.

## The gate this needs, which matters more than the one fix

Add a **connectivity test for every loss in `nn::loss`**, not just for `SmoothL1Loss`. The
shape of the defect is "the loss returns a plausible number and a dead graph", which no
value-based test can see — every existing test of this loss passes. One test per loss that
runs backward and asserts a non-zero gradient reaches `pred` is what makes the class of bug
visible. Without it, the next loss written this way lands the same way.

## Provenance

- Surfaced: Phase 6 (native time-series forecasting stack), spike 002 (NeuralProphet port).
- Deferred deliberately from Phase 6 as D-ITEM-06-02: fixing it means touching
  `aprender-core`, which pulls its full lib test suite into every cycle of a forecasting
  plan. Recorded in
  `.planning/phases/06-native-time-series-forecasting-stack/deferred-items.md`.

**Comment by noahgift** [comment]
Thanks!  Will fix in upcoming releases shortly!  Keep them coming.

**Comment by noahgift** [comment]
triage 2026-09-23 (aprender-dd, standing triage owner): milestoned to 0.75.0, theme CRUX fine-tune/distill (training); it moved out of `backlog` (the 2026-09-17 overflow bucket, not a release).
